### ISSUE TITLE
Enable __residual for "void" type

### ISSUE THREAD

**Issue by NTillmann** [facebook/prepack#621]
This is not yet support, but there's no deep reason as to why, and it's a bit annoying to work around this with a dummy result.

```js
__residual("void", function() { /* code with no visible side effects here */ });
```

**Comment by sebmarkbage** [comment]
We might want to clarify what "visible" side effects are. Since in this case it clearly will have some side-effect.

**Comment by NTillmann** [comment]
"Visible side effects" means side effects that would mutate any objects on the JavaScript heap.

**Comment by NTillmann** [comment]
To implement this, start by chasing down what happens in the definition of `__residual` in `src/intrinsics/prepack/global.js`, and simply omit the construction of a result value in the case of a `void` type. Either by supporting `void` directly in the called functions, or by creating new variations of the called functions specifically for the `void` case, if that results in cleaner code.

Add a test case in `test/serializer/abstract` that verifies that it works.

**Comment by sebmarkbage** [comment]
I think "visible" means more specifically that something else in the program after this residual will read from those things.

It is fine to mutate objects on the JavaScript heap, e.g. if this residual function is last in the program.

This example is not quite last in the program, and it will mutate the JS heap, but not the fields that are read immediately after this.

https://github.com/facebook/prepack/blob/master/src/prepack-cli.js#L155

**Comment by NTillmann** [comment]
> I think "visible" means more specifically that something else in the program after this residual will read from those things. It is fine to mutate objects on the JavaScript heap, e.g. if this residual function is last in the program.

You are right. My previous statement was too conservative.

**Comment by NTillmann** [comment]
Implemented by #709.
